testcases/kernel/syscalls/fcntl/fcntl27.c

```diff
@@ -1,192 +1,51 @@
-
-#include <sys/types.h>
-#include <sys/stat.h>
-#include <fcntl.h>
-#include <unistd.h>
-#include <errno.h>
-#include <string.h>
-#include <signal.h>
-#include "test.h"
-
-void setup();
-void cleanup();
-
-char *TCID = "fcntl27";
-int TST_TOTAL = 1;
-
-char fname[255];
-int fd;
-
-int main(int ac, char **av)
-{
-	int lc, expected_result = -1;
-    /***************************************************************
-     * parse standard options
-     ***************************************************************/
-	tst_parse_opts(ac, av, NULL, NULL);
-
-    /***************************************************************
-     * perform global setup for test
-     ***************************************************************/
-	setup();
-
-	expected_result = -1;
-
-    /***************************************************************
-     * check looping state if -c option given
-     ***************************************************************/
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
+ * Basic test for fcntl(2) using F_SETLEASE & F_RDLCK argument.
+ */
 
-		tst_count = 0;
+#include "lapi/fcntl.h"
+#include "tst_test.h"
 
 #ifdef F_SETLEASE
-		/*
-		 * Call fcntl(2) with F_SETLEASE & F_RDLCK argument on fname
-		 */
-		TEST(fcntl(fd, F_SETLEASE, F_RDLCK));
-
-		/* check return code */
-		if (TEST_RETURN == expected_result) {
-			tst_resm(TPASS,
-				 "fcntl(fd, F_SETLEASE, F_RDLCK) succeeded");
-		} else {
-			tst_resm(TFAIL, "fcntl(%s, F_SETLEASE, F_RDLCK)"
-				 " failed with errno %d : %s", fname,
-				 TEST_ERRNO, strerror(TEST_ERRNO));
-		}
-#else
-		tst_resm(TINFO, "F_SETLEASE not defined, skipping test");
-#endif
-	}
 
-    /***************************************************************
-     * cleanup and exit
-     ***************************************************************/
-	cleanup();
-	tst_exit();
-
-}
-
-/***************************************************************
- * setup() - performs all ONE TIME setup for this test.
- ***************************************************************/
-void setup(void)
+static struct test_case
 {
-
-	tst_sig(NOFORK, DEF_HANDLER, cleanup);
-
-	TEST_PAUSE;
-
-	tst_tmpdir();
-
-	sprintf(fname, "tfile_%d", getpid());
-	if ((fd = open(fname, O_RDWR | O_CREAT, 0777)) == -1) {
-		tst_brkm(TBROK, cleanup,
-			 "open(%s, O_RDWR|O_CREAT,0777) Failed, errno=%d : %s",
-			 fname, errno, strerror(errno));
-	}
-}
-
-/***************************************************************
- * cleanup() - performs all ONE TIME cleanup for this test at
- *				 completion or premature exit.
- ***************************************************************/
-void cleanup(void)
+	int oflags;
+	mode_t mode;
+	int cmd;
+	char *cmd_desc;
+	int arg;
+	char *arg_desc;
+	int exp_errno;
+} tcases[] = {
+	{O_RDWR | O_CREAT, 0777, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
+	{O_WRONLY | O_CREAT, 0222, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
```

Comment:
More compact output we often use:
```C
#define FLAG_DESC(x, y) .cmd = x, .cmd_desc = #x, .arg = y, .arg_desc = #y
// ...
{O_RDWR | O_CREAT, 0777, FLAG_DESC(F_SETLEASE, F_RDLCK)},
```

Reply:
Actually `EAGAIN`, `F_SETLEASE` and `F_RDLCK` are the same for both items. Why not keep just `oflags` and `mode` in `struct test_case` and the rest pass either directly or define if need to use on more places?

```C
static struct test_case
{
	int oflags;
	mode_t mode;
} tcases[] = {
	{O_RDWR | O_CREAT, 0777},
	{O_WRONLY | O_CREAT, 0222},
};
// ...
int fd = SAFE_OPEN("testfile", tc->oflags, tc->mode);
TST_EXP_FAIL(fcntl(fd, F_SETLEASE, F_RDLCK));

Reply:
It was the original version, then I followed @Avinesh advice :) 
https://github.com/linux-test-project/ltp/pull/1132/#discussion_r1497448720

Reply:
Thinking about it, I tend to agree with @pevik here. Keept it simple and if needed we can extend it later.

Reply:
@pevik , Thank you for suggesting the correct approach.
Sorry @ilmanzo for the creating confusion. :)